components/net/http_cache.rs

```diff
@@ -302,6 +297,185 @@ fn create_cached_response(request: &Request, cached_resource: &CachedResource, c
+                            let new_resource = create_resource_with_bytes_from_resource(&bytes, partial_resource);
+                            let cached_response = create_cached_response(request, &new_resource, &*headers);
+                            return Some(cached_response);
+                        }
+                    }
+                }
+            },
+            (&header::ByteRangeSpec::Last(offset), Some(ref complete_resource)) => {
+                if let ResponseBody::Done(ref body) = *complete_resource.body.lock().unwrap() {
+                    let from_byte = body.len() - offset as usize;
+                    let requested = body.get(from_byte..);
+                    if let Some(bytes) = requested {
+                        let new_resource = create_resource_with_bytes_from_resource(bytes, complete_resource);
+                        let cached_headers = new_resource.metadata.headers.lock().unwrap();
+                        let cached_response = create_cached_response(request, &new_resource, &*cached_headers);
+                        return Some(cached_response);
+                    }
+                }
+            },
+            (&header::ByteRangeSpec::Last(offset), None) => {
+                for partial_resource in partial_cached_resources {
+                    let headers = partial_resource.metadata.headers.lock().unwrap();
+                    let content_range = headers.get::<header::ContentRange>();
+                    let (res_beginning, res_end, total) = match content_range {
+                        Some(&header::ContentRange(
+                            header::ContentRangeSpec::Bytes {
+                                range: Some((res_beginning, res_end)),
+                                instance_length: Some(total) })) => (res_beginning, res_end, total),
+                        _ => continue,
+                    };
+                    if (total - res_beginning) > (offset - 1 ) && (total - res_end) < offset + 1 {
+                        let resource_body = &*partial_resource.body.lock().unwrap();
+                        let requested = match resource_body {
+                            &ResponseBody::Done(ref body) => {
+                                let from_byte = body.len() - offset as usize;
+                                body.get(from_byte..)
+                            },
+                            _ => continue,
+                        };
+                        if let Some(bytes) = requested {
+                            let new_resource = create_resource_with_bytes_from_resource(&bytes, partial_resource);
+                            let cached_response = create_cached_response(request, &new_resource, &*headers);
+                            return Some(cached_response);
+                        }
+                    }
+                }
+            }
+        }
+    } else {
+        // Not a Range request.
+        if let Some(ref cached_resource) = candidates.first() {
```

Comment:
I actually meant that the ranged check and non-ranged behavior could stay in the caller, just yo reduce the complexity of this function.